projects/ad713x_fmcz/ad713x_fmc.c

```diff
@@ -79,126 +88,170 @@
+	ad713x_init_param_2.gpio_pnd.number = GPIO_PDN_2;
+	ad713x_init_param_2.gpio_pnd.extra = (void *)&gpio_extra_param;
+	ad713x_init_param_2.gpio_resetn.number = GPIO_RESETN_2;
+	ad713x_init_param_2.gpio_resetn.extra = (void *)&gpio_extra_param;
+	ad713x_init_param_2.mode_master_nslave = false;
+	ad713x_init_param_2.dclkmode_free_ngated = false;
+	ad713x_init_param_2.dclkio_out_nin = false;
+	ad713x_init_param_2.pnd = true;
+	ad713x_init_param_2.spi_init_prm.chip_select = AD7134_2_SPI_CS;
+	ad713x_init_param_2.spi_init_prm.max_speed_hz = 100000000;
+	ad713x_init_param_2.spi_init_prm.mode = SPI_MODE_3;
+	ad713x_init_param_2.spi_init_prm.extra = (void *)&spi_ps_extra;
+	ad713x_init_param_2.spi_common_dev = 0;
 
-	spi_eng_msg_cmds[0] = SLEEP(2000); // 2us
-	spi_eng_msg_cmds[1] = TRANSFER_BYTES_R(3);
+	spi_eng_msg_cmds[0] = SLEEP(100);
+	spi_eng_msg_cmds[1] = TRANSFER_BYTES_R(1);
 
 	Xil_ICacheEnable();
 	Xil_DCacheEnable();
 
-	ad713x_init(&dev, ad713x_default_init_param);
+	ad713x_init(&ad713x_dev_1, &ad713x_init_param_1);
 
-	spi_eng_init(&desc, spi_eng_init_params);
+	mdelay(1000);
 
-	ad713x_dig_filter_sel_ch(dev, SINC3, CH0);
+	ad713x_init_param_2.spi_common_dev = ad713x_dev_1->spi_desc;
 
-	/* For this example, only offload is supported*/
-	msg = (spi_eng_msg *)malloc(sizeof(*msg));
+	ad713x_init(&ad713x_dev_2, &ad713x_init_param_2);
+
+	mdelay(1000);
+
+	spi_eng_offload_init(&spi_eng_desc, &spi_eng_inits);
+
+	/** For this example, only offload is supported */
+	msg = calloc(1, sizeof *msg);
 	if (!msg)
 		return -1;
 
+	spi_eng_set_transfer_length(spi_eng_desc, 32);
+
 	msg->spi_msg_cmds = malloc(sizeof(spi_eng_msg_cmds));
 	msg->spi_msg_cmds = spi_eng_msg_cmds;
 	msg->msg_cmd_len = ARRAY_SIZE(spi_eng_msg_cmds);
 	msg->rx_buf_addr = 0x800000;
 	msg->tx_buf_addr = 0xA000000;
+	msg->rx_buf = calloc((AD7134_FMC_CH_NO * AD7134_FMC_SAMPLE_NO),
+			     sizeof(uint32_t));
 
-	/* Init the rx buffer with 0s */
-	for (i = 0; i < ARRAY_SIZE(msg->rx_buf); i++)
-		msg->rx_buf[i] = 0;
+	spi_eng_offload_load_msg(spi_eng_desc, msg);
+	spi_eng_transfer_multiple_msgs(spi_eng_desc,
+				       (AD7134_FMC_CH_NO * AD7134_FMC_SAMPLE_NO));
 
-	spi_eng_offload_load_msg(desc, msg);
-	spi_eng_transfer_multiple_msgs(desc, 1024);
+	mdelay(1000);
 
-	mdelay(10000);
+	Xil_DCacheInvalidateRange(0x800000, 16384 * 16);
 
 	free(msg);
 
-	offload_data = (uint32_t*)desc->rx_dma_startaddr;
-	for(i = 0; i < (desc->rx_length/4); i++) {
-		printf("CH%lu: 0x%lx\r\n", i%8, *offload_data);
-		offload_data += 1; // go to the next address in memory
+	const float lsb = 4.096 / (pow(2, 23));
+	float data;
+	uint8_t j;
+
+	offload_data = (uint32_t*)((struct spi_desc_extra *)spi_eng_desc->extra)->
+		       rx_dma_startaddr;
+	for(i = 0; i < AD7134_FMC_SAMPLE_NO; i++) {
+		j = 0;
+		while(j < 8) {
+			*(offload_data + j) <<= 1;
+			*(offload_data + j) &= 0xffffff00;
+			*(offload_data + j) >>= 8;
+
+			data = lsb * ((int32_t)*(offload_data + j) & 0xFFFFFF);
```

Comment:
Replace:
`data = lsb * ((int32_t)*(offload_data + j) & 0xFFFFFF);`
for better understanding, with:
`data = (*(offload_data + j) & 0xFFFFFF) >> 10;`

Reply:
Is this the same equation? It seems that it would introduce a full scale error of 4-4.096=96mV